includes/payment-methods/class-upe-payment-gateway.php

```diff
@@ -418,21 +437,37 @@ public function process_redirect_payment( $order_id, $intent_id, $save_payment_m
 			// Get user/customer for order.
 			list( $user, $customer_id ) = $this->manage_customer_details_for_order( $order );
 
+			$payment_needed = 0 < $order->get_total();
+
 			// Get payment intent to confirm status.
-			$intent                 = $this->payments_api_client->get_intent( $intent_id );
-			$status                 = $intent->get_status();
-			$charge_id              = $intent->get_charge_id();
-			$currency               = $intent->get_currency();
-			$payment_method_id      = $intent->get_payment_method_id();
-			$payment_method_details = $intent->get_payment_method_details();
-			$payment_method_type    = $payment_method_details['type'];
+			if ( $payment_needed ) {
+				$intent                 = $this->payments_api_client->get_intent( $intent_id );
+				$client_secret          = $intent->get_client_secret();
+				$status                 = $intent->get_status();
+				$charge_id              = $intent->get_charge_id();
+				$currency               = $intent->get_currency();
+				$payment_method_id      = $intent->get_payment_method_id();
+				$payment_method_details = $intent->get_payment_method_details();
+				$payment_method_type    = $payment_method_details['type'];
+				$error                  = $intent->get_last_payment_error();
+			} else {
+				$intent                 = $this->payments_api_client->get_setup_intent( $intent_id );
+				$client_secret          = $intent['client_secret'];
+				$status                 = $intent['status'];
+				$charge_id              = '';
+				$currency               = $order->get_currency();
+				$payment_method_id      = $intent['payment_method'];
+				$payment_method_details = false;
+				$payment_method_options = array_keys( $intent['payment_method_options'] );
+				$payment_method_type    = $payment_method_options[0];
+				$error                  = $intent['last_setup_error'];
+			}
```

Comment:
This slightly ugly split logic here is similar to [this logic](https://github.com/Automattic/woocommerce-payments/blob/develop/includes/class-wc-payment-gateway-wcpay.php#L955-L980) here in the old WCPay card gateway. The reason for it is because the `WC_Payments_API_Client::get_intent` function returns a `WC_Payments_API_Intention` object, whereas `WC_Payments_API_Client::get_setup_intent` simply returns an array representing the `WP_REST_Response` sent back from the Payments Server. I wanted to refactor this to make this a bit cleaner, by making it possible to deserialise both REST responses into a `WC_Payments_API_Intention` object: I don't think this should be too tricky, since the only real difference is that the Setup Intent lacks a currency field and any accompanying charge objects. However, this refactor would involve a bit of additional work on the old gateway and a few other places in the codebase, pretty irrelevant to this PR, so I created #2422 instead and decided to save this adventure for another day.